src/Key.php

```diff
@@ -5,18 +5,132 @@
 class Key
 {
     /** @var string */
-    public $id;
+    private $id;
 
     /** @var string */
-    public $secret;
+    private $secret;
+
+    /** @var string */
+    private $certificate;
+
+    /** @var string */
+    private $privateKey;
+
+    /** @var string */
+    private $type;
 
     /**
      * @param string $id
      * @param string $secret
      */
-    public function __construct($id, $secret)
+    public function __construct($id, $item)
     {
         $this->id = $id;
-        $this->secret = $secret;
+        $certificate = $this->getX509Certificate($item);
+        $privateKey = $this->getRSAPrivateKey($item);
+        if (($certificate || $privateKey)) {
+            $this->type = 'rsa';
+            if ($privateKey) {
+                $this->privateKey = $privateKey;
+            }
+            if ($certificate) {
+                $this->certificate = openssl_x509_read($certificate);
+            }
+            if ($certificate && $privateKey) {
+                if (!openssl_x509_check_private_key(
+                        $this->certificate,
+                        $this->privateKey
+                    )
+                ) {
+                    throw new KeyException('Supplied Certificate and Key are not related');
+                }
+            }
+        } else {
+            $this->type = 'secret';
+            $this->secret = $item;
+            $publicKey = null;
+            $privateKey = null;
+        }
+    }
+
+    private function getRSAPrivateKey($object)
+    {
+        if (is_array($object)) {
```

Comment:
Under what circumstances is this going to be an array and why? Ideally avoid allowing multiple types in arguments

Reply:
The user either supplies:
- A Public Key/Certificate for verifying as string/object
- A Private Key for signing as string/object
- Or both together in an array of string/objects - this can now be used for both signing and verifying without having to construct different objects for each case.

Allowing both forms means we can use the same KeyStore object for all three cases to derive a Key object. Alternative is to expose HMAC and Asymmetric-specific Key interfaces which is both a BC and requires more documentation and user education for little gain.

Reply:
@mtibben Is this ok? I realise the input here is fungible, but I'd rather that than expect the user to do key/certificate handling to a more advanced degree.